### PR TITLE
`static operator()` for stateless functors

### LINKED ISSUE #4130: Many call operators (`operator()`) can be `static` when `defined(__cpp_static_call_operator)`
Clang 16 supports `static operator()` and defines `__cpp_static_call_operator` in all of our supported language modes, with a suppressible warning "before" C++23. (MSVC will eventually provide similar support in all language modes.) Static and non-static `operator()` have different mangled names, so we can simultaneously support both even in different TUs linked together. We need to audit our declarations of `operator()` and decide which can be declared `static`.

I think we'll want to do this by adding a couple of macro definitions:
1. a macro that expands to `static` when `defined(__cpp_static_call_operator)` and an empty string otherwise, and
2. a macro that expands to an empty string when `defined(__cpp_static_call_operator)` and 'const' otherwise.

We can then change the pertinent `operator()` declarations from e.g.:
```c++
constexpr int operator()(int x) const {
    return x + 1;
}
```
to
```c++
_FIRSTMACRO constexpr int operator()(int x) _SECONDMACRO {
    return x + 1;
}
```

I've started to prototype this in https://github.com/microsoft/STL/commit/d85b96c45dfb3689818a4abdbe5e191f1cd8a818; feel free to use that as a starting point. (I'm not fond of the names I chose for the macros, feel free to change them.)

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #4130. This PR makes the `operator()`s of stateless functors `static`, including those of captureless lambdas, except for...
- some `operator()`s that are intentionally not `const` to me (e.g. `_Rand_urng_from_func::operator()`),
- the `operator()`s that explicitly specified `const` and non-`static` in the standard wording (except for that of _`synth-three-way`_ where the functor itself is exposition-only),
  - I tried submitting an LWG issue for them, but Daniel told me to write a proposal for changing; and
- the lambda expression used in `basic_format_arg::handle`, because there seems a bug of Clang 17 involving calling conventions for static lambdas, which seems fixed in Clang 18.

<details><summary>List of explicitly <code>const</code> and non-<code>static</code> functors in the standard library</summary>
<p>

`<memory>`:
- `std::default_delete`
- `std::owner_less`
- `std::owner_hash`
- `std::owner_equal`

`<type_traits>`:
- `std::integral_constant`

`<chrono>`:
- `std::chrono::clock_time_conversion`

`<functional>`:
- `std::plus`
- `std::minus`
- `std::multiplies`
- `std::divides`
- `std::modulus`
- `std::negate`
- `std::equal_to`
- `std::not_equal_to`
- `std::greater`
- `std::less`
- `std::greater_equal`
- `std::less_equal`
- `std::compare_three_way`
- `std::ranges::equal_to`
- `std::ranges::not_equal_to`
- `std::ranges::greater`
- `std::ranges::less`
- `std::ranges::greater_equal`
- `std::ranges::less_equal`
- `std::logical_and`
- `std::logical_or`
- `std::logical_not`
- `std::bit_and`
- `std::bit_or`
- `std::bit_xor`
- `std::bit_not`
- `std::identity`

</p>
</details>

Drive-by changes:
- Change `(*this)(...)` to `operator()(...)` whenever `operator()` is possibly `static`.
  - `ranges::swap` needs some special handling.
- Make some internal helper member functions of fold algorithms to `static`.
- Make `_Cmp_cs::operator()` conditionally `const` or `static` - it seems unintended to me that it wasn't `const`.
- Slightly reform some parallel memory algorithms to instantiate less functor types when the iterator type varies.

### CODE REVIEW

File: stl/inc/format [hunk 1]
```diff
@@ -701,8 +701,8 @@ public:
     public:
         template <class _Ty>
         explicit handle(_Ty& _Val) noexcept
-            : _Ptr(_STD addressof(_Val)),
-              _Format([](basic_format_parse_context<_CharType>& _Parse_ctx, _Context& _Format_ctx, const void* _Ptr) {
+            : _Ptr(_STD addressof(_Val)), _Format([](basic_format_parse_context<_CharType>& _Parse_ctx,
+                                                      _Context& _Format_ctx, const void* _Ptr) _STATIC_CALL_OPERATOR {
```
Comment (StephanTLavavej) [inline stl/inc/format:705]: There are x86 test failures here:

```
D:\build\out\inc\format(704,51): error: cannot initialize return object of type 'void (*)(basic_format_parse_context<_CharType> &, std::basic_format_context<std::back_insert_iterator<std::_Fmt_buffer<char>>, char> &, const void *)' (aka 'void (*)(basic_format_parse_context<char> &, std::basic_format_context<std::back_insert_iterator<std::_Fmt_buffer<char>>, char> &, const void *)') with an lvalue of type 'void (basic_format_parse_context<_CharType> &, std::basic_format_context<std::back_insert_iterator<std::_Fmt_buffer<char>>, char> &, const void *) __attribute__((thiscall))' (aka 'void (basic_format_parse_context<char> &, std::basic_format_context<std::back_insert_iterator<std::_Fmt_buffer<char>>, char> &, const void *) __attribute__((thiscall))')
  704 |             : _Ptr(_STD addressof(_Val)), _Format([](basic_format_parse_context<_CharType>& _Parse_ctx,
      |                                                   ^
D:\build\out\inc\format(2062,62): note: in instantiation of function template specialization 'std::basic_format_arg<std::basic_format_context<std::back_insert_iterator<std::_Fmt_buffer<char>>, char>>::handle::handle<const std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<long long, std::ratio<1, 1000000>>>>' requested here
 2062 |             _Store_impl<_Erased_type>(_Arg_index, _Arg_type, static_cast<_Erased_type>(_Val));
```

`_Format` is a `__cdecl` function pointer, which is why this is x86-specific:

https://github.com/microsoft/STL/blob/202e3820d6390505855ce9a58a4bcee4f28feacd/stl/inc/format#L698

Presumably something about the `static` function call operator is interfering with the conversion from the lambda here. I suspect that Clang may not have considered the "omniconvertible" behavior (that MSVC originally implemented) when they added `static operator()`.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/format (hunk 1 above)
Comment (frederick-vs-ja) [reply stl/inc/format:705]: This seems fixed in Clang 18 ([Godbolt link](https://godbolt.org/z/6zbPqo89d), despite that Clang-cl is not properly supported on Godbolt).

It seems that we should drop `_STATIC_CALL_OPERATOR` at this moment.

File: stl/inc/concepts [hunk 2]
```diff
@@ -122,26 +122,26 @@ namespace ranges {
         struct _Cpo {
             template <class _Ty1, class _Ty2>
                 requires _Use_ADL_swap<_Ty1, _Ty2>
-            constexpr void operator()(_Ty1&& __t, _Ty2&& __u) const
+            constexpr _STATIC_CALL_OPERATOR void operator()(_Ty1&& __t, _Ty2&& __u) _CONST_CALL_OPERATOR
```
Comment (StephanTLavavej) [inline stl/inc/concepts:125]: `constexpr _STATIC_CALL_OPERATOR` is the opposite of our conventional order. Occurs 3 times in this file.

File: stl/inc/xutility [hunk 3]
```diff
@@ -987,7 +988,8 @@ namespace ranges {
         public:
             template <class _Ty1, class _Ty2>
                 requires (_Choice<_Ty1, _Ty2>._Strategy != _St::_None)
-            constexpr void operator()(_Ty1&& _Val1, _Ty2&& _Val2) const noexcept(_Choice<_Ty1, _Ty2>._No_throw) {
+            constexpr _STATIC_CALL_OPERATOR void operator()(_Ty1&& _Val1, _Ty2&& _Val2) _CONST_CALL_OPERATOR
```
Comment (StephanTLavavej) [inline stl/inc/xutility:991]: One more occurrence of `constexpr _STATIC_CALL_OPERATOR` contrary to our conventional order.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for this surprisingly large PR - I always forgot how big the STL has grown! :joy_cat:

I pushed a few changes, including a couple of refactorings - please double-check. I've also relabeled this as a (minor) performance improvement.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for powering the STL with `static` electricity! :zap: :zap: :zap:

**Pull request merged**